enhancements/cluster-high-availability-mode-api.md

```diff
@@ -0,0 +1,360 @@
+
+### Use the cluster profile to have different Deployments for operators
+
+We could use different manifests to have the cluster-version-operator
+deploy operators with different configuration (via command line
+arguments or environment variables). This would give us a way to
+configure operators differently for each profile, but those settings
+would be less discoverable and we would not be able to offer users a
+way to change them at runtime (if we ever want to do that).
+
+### Recognize HA/non-HA configuration based on control plane node count
+
+Operators could look at the number of control plane nodes and use that
+information to determine whether HA is enabled (3 or more nodes) or
+not (1 node).
+
+This would require giving the operator read access to Node resources,
+and would be subject to race conditions as the control plane
+forms. For example, the cluster-etcd-operator [requires 3 control
+plane
+nodes](https://github.com/openshift/cluster-etcd-operator/blob/98590e6ecfe282735c4eff01432ae40b29f81202/pkg/etcdenvvar/etcd_env.go#L72)
+to exist before deploying its operand. If it counted nodes, it could
+not tell the difference between a cluster starting to form with 1
+node, then more, or a cluster that is only ever going to include 1
+node.
+
+Looking at nodes also will not work in clusters using an external
+control plane, since there will not be any special control plane
+nodes.
+
+### Use operator-specific APIs to manage the behaviors
+
+The cluster-etcd-operator and cluster-authentication-operator both
+include flags to enter an unsupported non-HA operating mode. We could
+change those APIs to make the flags supported, and add similar flags
+to other components, such as cluster-ingress-operator and
+machine-config-operator. This would offer more flexibility, since each
+component could be configured separately. However, these settings are
+needed during cluster bootstrapping, so this approach would tie the
+installer to the APIs for each of the operators so they could be
+configured separately.
+
+Complexity will grow further if we add capabilities for managing the
+resource consumption of operators and their operands, since we risk
+having different ways to express those constraints in different APIs.
+
+### Manage upgrades by tracking settings in each consuming operator
+
+When we make the API mutable we will need to protect against older
+components encountering unknown enum values during an upgrade.
+Anything that use the `capabilities` API could copy the settings into
+its own API resource. If the enum value is unknown, it wouldn't be
+copied and the existing configuration would remain in place until the
+newer version of the component is deployed. This approach would
+require similar logic in each component, which wouldn't be complicated
+to implement but which would nonetheless be easy to overlook when
+implementing a feature.
+
+### Configure OLM-managed operators without the Infrastructure API
```

Comment:
@mrunalp @ecordell please look at this section, too

Reply:
I think @ecordell mentioned that there is already an api that the operators can access. Maybe OLM can query Infrastructure and fill it?

Reply:
That's not what we discussed today. See option 2 on line 354.